### ISSUE TITLE
Partial Pausability

### ISSUE THREAD

**Issue by wiasliaw** [solidstate-network/solidstate-solidity#153]
# Motivation

## a.

`PausableStorage` uses a full slot but only use 1 byte for boolean.

```solidity
struct Layout {
    bool paused;
}
```

## b.

When it comes to diamond's facets, partial pausability might in need.

For example,
- facet A is not associated with facet B, pause A should not pause B.
- facet B is associated with facet C, pause B should also pause C.

# Rationale

`PausableStorage` should be a uint256 which represent a bitmap `mapping(uint8 => bool)`. Each facet can pick a value(uint8) for pausability and not to bother with other facets.

```solidity
struct Layout {
    uint256 paused;
}
```

# Example

- pause A but not pause B
- pause B also pause C
- combining A and BC, pause D also pause A, B, C

```solidity
contract FacetA {
    uint256 private constant mask = 1 << 0; // bitmap[0]
    function foo1() whenNotPause(mask) { ... }
}

contract FacetB {
    uint256 private constant mask = 1 << 1; // bitmap[1]
    function foo2() whenNotPause(mask) { ... }
}

contract FacetC {
    uint256 private constant mask = 1 << 1; // bitmap[1]
    function foo3() whenNotPause(mask) { ... }
}

contract FacetD {
    uint256 private constant mask = 1 << 1 || 1 << 0; // bitmap[1] || bitmap[0]
    function foo4() whenNotPause(mask) { ... }
}
```

**Comment by ItsNickBarry** [comment]
This is a good idea.  As long as we keep an interface for pausing without supplying a parameter, then I would be in favor of merging this.

Maybe slot 255 could be used as a default.

**Comment by wiasliaw** [comment]
How about define an unimplement function to custom which bit should be use?

proposal here: https://github.com/wiasliaw/solidstate-pausable

**Comment by ItsNickBarry** [comment]
That wouldn't work with the "visibility layers" pattern.  The recommended way to share internal logic between facets is to share a single internal contract between multiple facets.  Using different versions of the same internal code would be confusing.  Your previous idea described above would work better.

**Comment by wiasliaw** [comment]
The second idea is to meet "keep an interface for pausing without supplying a parameter". I don't think it should keep the interface always the same. Additionally, the interface for external function hasn't altered, only alter the internal functions.

i.e.
```solidity
contract FacetA is Pausable {
    uint256 private constant mask = 1 << 0; // bitmap[0]

    function foo1() whenNotPause(mask) { ... }

    function pause() external {
        _pause(mask);
    }
}
```

**Comment by ItsNickBarry** [comment]
What I meant was something like this:
```solidity
modifier whenNotPaused() {
  ...
}

modifier whenNotPaused(uint8 mask) {
  ...
}
```

But I see that this is not possible, because modifiers cannot be overloaded.  Instead, let's keep the existing modifiers such as `whenNotPaused()`, and add new ones like `whenNotPartiallyPaused(uint8 mask)`.

**Comment by ItsNickBarry** [comment]
@iam-dev Yeah, that's fine, we can use a separate `paused` variable for non-partial pauses.

**Comment by wiasliaw** [comment]
I dont feel well these days..

@ItsNickBarry OK, I will work on proposal repo.

---

@iam-dev 

I dont like the storage be this. One of the reason to address this issue is that boolean with unstruct storage will occupy a full slot but only use one bit. Mappging type with boolean works the same.

Otherwise, it is not capable with facets in diamond pattern. According your desgin, it should check in diamond's fallback function before delegatecall.

**Comment by ItsNickBarry** [comment]
The code from this PR has been included in #199 and modified in such a way that it does not work as originally intended.  Published in `v0.0.55`.  Thanks @iam-dev for the contribution.

**Comment by ItsNickBarry** [comment]
The code from this PR has been included in #199 and modified in such a way that it does not work as originally intended.  Published in `v0.0.55`.  Thanks @wiasliaw for the contribution.
